### PR TITLE
Inline oc.js when possible!

### PR DESCRIPTION
**Pull request by rullzer**
With https://github.com/nextcloud/server/pull/1871 in we can now do fancy pancy stuff. Like inline javascript safely.

This PR bring you the inlining of oc.js. This file was loaded on each pageload separatly and was a blocking call! Now if your browser support CSPv3 it will inline the file saving you a blocking (non cacheable request!).
- Move oc.js generation to helper class
- Move CSPv3 test
- Inline oc.js when possible

@LukasReschke as discussed!

CC: @MorrisJobke @nickvergessen @icewind1991

### CODE REVIEW

File: core/Controller/OCJSController.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use OCP\IRequest;
+use OCP\IURLGenerator;
+use OCP\IUserSession;
+
+class OCJSController extends Controller {
+
+	/** @var JSConfigHelper */
+	private $helper;
+
+	public function __construct($appName,
```
Comment (LukasReschke) [inline core/Controller/OCJSController.php]: A kingdom for PHPDocs 🚀

File: core/templates/layout.base.php [hunk 2]
```diff
@@ -18,6 +18,11 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php if (isset($_['inline_ocjs'])): ?>
+			<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
+				<?php echo($_['inline_ocjs']); ?>
```
Comment (LukasReschke) [inline core/templates/layout.base.php]: `print_unescaped`? :)

File: core/templates/layout.guest.php [hunk 3]
```diff
@@ -19,6 +19,11 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php if (isset($_['inline_ocjs'])): ?>
+			<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
+				<?php echo($_['inline_ocjs']); ?>
```
Comment (LukasReschke) [inline core/templates/layout.guest.php]: `print_unescaped`? :)

File: core/templates/layout.user.php [hunk 4]
```diff
@@ -26,6 +26,11 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php if (isset($_['inline_ocjs'])): ?>
+			<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
+				<?php echo($_['inline_ocjs']); ?>
```
Comment (LukasReschke) [inline core/templates/layout.user.php]: `print_unescaped`? :)

File: lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php [hunk 5]
```diff
@@ -102,7 +105,8 @@ public function __construct(IRequest $request,
 								$isLoggedIn,
 								$isAdminUser,
 								ContentSecurityPolicyManager $contentSecurityPolicyManager,
-								CsrfTokenManager $csrfTokenManager) {
+								CsrfTokenManager $csrfTokenManager,
+								ContentSecurityPolicyNonceManager $cspNonceManager) {
```
Comment (LukasReschke) [inline lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php]: Update PHPDoc? ;)

**Review by LukasReschke — CHANGES_REQUESTED** [review]
Awesome! Some nitpicks are still existent though ;)

**Comment by rullzer** [comment]
@LukasReschke fixed

**Review by LukasReschke — APPROVED** [review]

**Comment by LukasReschke** [comment]
Good stuff! :+1:

**Comment by MorrisJobke** [comment]
Code looks good and I tested this in Firefox, Safari, Chrome, IE and Edge.

**Comment by LukasReschke** [comment]
Small addition: https://github.com/nextcloud/server/pull/1920 :)

**Comment by nickvergessen** [comment]
Please also delete the old file, just cost björn and me some minutes to figure out, why the break point didn't stop...

**Pull request merged**